doc:Add to block=storage and create alert-manager configuration docum… (#4114)
cortexproject/cortex: docs/guides/alert-manager-configuration.md

---

> title: "Configuring Notification using Cortex AlertManager"

Reviewer: The alertmanager is spelled "Alertmanager" in Prometheus ([see here](https://prometheus.io/docs/alerting/latest/alertmanager/)) so I would change "AlertManger" to "Alertmanager" across this PR to keep it consistent with Prometheus.

linkTitle: "AlertManager configuration"
weight: 10

> slug: AlertManager-configuration

Reviewer: The slug should be all lowercase.

```suggestion
slug: alertmanager-configuration
```

Author: Done

---

### Context

Cortex AlertManager notification setup follow mostly the syntax of Prometheus AlertManager since it is based on the same codebase.  The following is a description on how to load the configuration setup so that AlertManager can use for notification when an alert event happened.

#### Cortex AlertManager configuration

> Cortex AlertManager can be uploaded via [Cortex Set Alertmanager API](https://cortexmetrics.io/docs/api/#set-alertmanager-configuration) or using Grafana Labs [Cortex Tools](https://github.com/grafana/cortex-tools).

Reviewer: Links should be relative in the doc. Something like this:

```suggestion
Cortex AlertManager can be uploaded via [Cortex Set Alertmanager API](../api/_index.md#set-alertmanager-configuration) or using Grafana Labs [Cortex Tools](https://github.com/grafana/cortex-tools).
```

Follow the instruction at the `cortextool` link above to download or update to the latest version of the tool.

To obtain the full help of how to use `cortextool` for all commands and flags, use
`cortextool --help-long`.

The following example shows the steps to upload the configuration to Cortex `alertmanager` using `cortextool`. 

1. Create `.yaml` file to specify the configuration for AlertManager notification characteristic.  

The following is `amconfig.yml`, an example of a configuration for Cortex `alertmanager` to send notification to a Slack channel:

```

> global:

Reviewer: I'm wondering if we should use a simpler alertmanager config as an example. The purpose of this guide is not to show how Alertmanager works, but how to configure it in Cortex. I've the feeling such a long example config may deviate the focus from what we want to show in this guide.

For example, the following in a SMTP-based receiver:

```
global:
  # The smarthost and SMTP sender used for mail notifications.
  smtp_smarthost: 'localhost:25'
  smtp_from: 'alertmanager@example.org'
  smtp_auth_username: 'alertmanager'
  smtp_auth_password: 'password'

route:
  # A default receiver.
  receiver: send-email

receivers:
  - name: send-email
    email_configs:
      - to: 'someone@localhost'
```

Author: Precisely my issue that motivates me to provide a more "real" example.  Lots of use cases are slack based and I spent countless hours trying to find a more "real" life setup to help with my application of Cortex.  So by providing a little more 'real" example,  I hope to help the users sometimes to comb through many many messages and issued to find the answer

Reviewer: I understand this, but to me it looks more like a topic for improving the Prometheus alertmanager doc.

Author: That's fine by me.
I got my answer :-)

              *Details:*
                {{ range .Labels.SortedPairs }} • *{{ .Name }}:* `{{ .Value }}`
                {{ end }}
              {{ end }}
```

[Example on how to setup Slack](https://grafana.com/blog/2020/02/25/step-by-step-guide-to-setting-up-prometheus-alertmanager-with-slack-pagerduty-and-gmail/#:~:text=To%20set%20up%20alerting%20in,to%20receive%20notifications%20from%20Alertmanager.) to support receiving AlertManager notification.

2. With your Cortex ID,  URL and API key,  you are ready to upload the configuration Alertmanager.

In this example,  Cortex `alertmanager` is set to be available via localhost on port 8095 with only one user/org = 0.

To upload the above configuration `.yml` file with `--key` to be your Basic Authentication or API key:

```
cortextool alertmanager load ./amconfig.yml \
--address=http://localhost:8095 \
--id=0
--key=<yourKey>
```
If there is no error reported,  the upload is successful.

3. To confirm that the configuration is uploaded correctly:

```
cortextool alertmanager get \
--adress=http://localhost:8095 \
--id=0
--key=<yourKey>
```

> To upload the configuration for Cortex `alertmanager` using Cortex API and curl - see [Cortex Set Alertmanager API](https://cortexmetrics.io/docs/api/#set-alertmanager-configuration).

Reviewer: ```suggestion
To upload the configuration for Cortex `alertmanager` using Cortex API and curl - see [Cortex Set Alertmanager API](../api/_index.md#set-alertmanager-configuration).
```

Author: Done

---
title: "Configuring Notification using Cortex AlertManager"
linkTitle: "AlertManager configuration"
weight: 10
slug: AlertManager-configuration
---

### Context

Cortex AlertManager notification setup follow mostly the syntax of Prometheus AlertManager since it is based on the same codebase.  The following is a description on how to load the configuration setup so that AlertManager can use for notification when an alert event happened

#### Cortex AlertManager configuration

Cortex AlertManager can be uploaded via [Cortex Set Alertmanager API](https://cortexmetrics.io/docs/api/#set-alertmanager-configuration) or using Grafana Labs [Cortex Tools](https://github.com/grafana/cortex-tools)

> The following describe the step to load the configuration to Cortex AlertManager using `cortextool`.  Follow the instruction above to download or update to the latest version of `cortextool`.

Reviewer 2: > The following describe the step ...

Suggestion: "The following example shows the steps to load ...".

> AlertManager

We should not capitalize "Manager". 

> Follow the instruction above ...

There was no instruction, only link to Cortex Tool.

To obtain the full help of how to use `cortextool` for all commands and flags, use
`cortextool --help-long`

> >  1. Create `.yaml` file to specify the configuration for AlertManager notification characteristic.  

Reviewer 2: This creates block quote. Perhaps we should use sub-headings instead?

> >>> The following is `amconfig.yml`, an example of a configuration for AlertManager to send notification to a Slack channel

Reviewer 2: Let's remove `>>>`, and make it a normal text instead of triple-blockquote. Same for the text below.

            text: >-
              {{ range .Alerts -}}
              *Alert:* {{ .Annotations.title }}
              *Description:* {{ .Annotations.description }}
                 • *Environment:* {{ if .Labels.environment }}`{{ .Labels.environment }}`{{ else }}`N/A`{{ end }}
                 • *Hostname:* {{ if .Labels.node }}`{{ .Labels.node }}`{{ else if .Labels.nodename }}`{{ .Labels.nodename }}`{{ else }}`N/A`{{ end }}
                 • *Condition:* {{ if .Labels.condition }}`{{ .Labels.condition }}`{{ else }}`N/A`{{ end }}
                 • *Current value:* {{ if .Annotations.current_value }}`{{ .Annotations.current_value }}`{{ else }}`N/A`{{ end }}
                 • *Opened at:* {{ if .StartsAt }}`{{ .StartsAt }}`{{ else }}`N/A`{{ end }}
                 • *Closed at:* {{ if eq .Status "resolved" }}`{{ .EndsAt }}`{{ else }}`Unresolved, alert still active.`{{ end }}
              *Details:*
                {{ range .Labels.SortedPairs }} • *{{ .Name }}:* `{{ .Value }}`
                {{ end }}
              {{ end }}
```

>>> [Full Example of Alertmanager configuration](https://github.com/prometheus/alertmanager/blob/master/doc/examples/simple.yml)

> >>> [Example on how to setup Slack](https://grafana.com/blog/2020/02/25/step-by-step-guide-to-setting-up-prometheus-alertmanager-with-slack-pagerduty-and-gmail/#:~:text=To%20set%20up%20alerting%20in,to%20receive%20notifications%20from%20Alertmanager.) to support receiving AlertManager notification

Reviewer 2: Let's simplify the link to `https://grafana.com/blog/2020/02/25/step-by-step-guide-to-setting-up-prometheus-alertmanager-with-slack-pagerduty-and-gmail` only.

> > 2. With your Cortex ID,  URL and API key,  you are ready to upload the configuration Alertmanager

Reviewer 2: Subheading.

>>> In this example,  AlertManager is set to be available via localhost on port 8095 with only one user/org = 0 

>>> To upload the above configuration `.yml` file with `--key` to be your Basic Authentication or API key

```
cortextool alertmanager load ./amconfig.yml \
--address=http://localhost:8095 \
--id=0
--key=<yourKey>
```
>>> If there is no error reported,  the upload is successful

> > 3. To confirm that the configuration is uploaded correctly

Reviewer 2: Subheading.

To obtain the full help of how to use `cortextool` for all commands and flags, use
`cortextool --help-long`.

The following example shows the steps to upload the configuration to Cortex `Alertmanager` using `cortextool`. 
#### 1. Create the Alertmanager configuration `yml` file.  

The following is `amconfig.yml`, an example of a configuration for Cortex `Alertmanager` to send notification via email:

```
global:
  # The smarthost and SMTP sender used for mail notifications.
  smtp_smarthost: 'localhost:25'
  smtp_from: 'alertmanager@example.org'
  smtp_auth_username: 'alertmanager'
  smtp_auth_password: 'password'

route:
  # A default receiver.
  receiver: send-email

receivers:
  - name: send-email
    email_configs:
      - to: 'someone@localhost'
```

[Example on how to setup Slack](https://grafana.com/blog/2020/02/25/step-by-step-guide-to-setting-up-prometheus-alertmanager-with-slack-pagerduty-and-gmail/#:~:text=To%20set%20up%20alerting%20in,to%20receive%20notifications%20from%20Alertmanager.) to support receiving Alertmanager notification.

#### 2. Upload the Alertmanager configuration

> In this example,  Cortex `Alertmanager` is set to be available via localhost on port 8095 with only one user/org = 0.

Author: ok.  I gave it the ID 100 instead

To upload the above configuration `.yml` file with `--key` to be your Basic Authentication or API key:

```
cortextool alertmanager load ./amconfig.yml \
--address=http://localhost:8095 \

> --id=0

Author: Done

--key=<yourKey>
```
If there is no error reported, the upload is successful.

To upload the configuration for Cortex `Alertmanager` using Cortex API and curl - see Cortex [Set Alertmanager configuration API](https://cortexmetrics.io/docs/api/#set-alertmanager-configuration).

#### 3. Ensure the configuration has been uploaded sucessfully

```
cortextool alertmanager get \
--adress=http://localhost:8095 \

> --id=0

Reviewer 2: missing `\` at the end of line

Author: Fixed

---
title: "Configuring Notification using Cortex Alertmanager"
linkTitle: "Alertmanager configuration"
weight: 10
slug: alertmanager-configuration
---

> ### Context

Reviewer: ```suggestion
## Context
```

Cortex Alertmanager notification setup follow mostly the syntax of Prometheus Alertmanager since it is based on the same codebase.  The following is a description on how to load the configuration setup so that Alertmanager can use for notification when an alert event happened.

> #### Cortex Alertmanager configuration

Reviewer: ```suggestion
### Cortex Alertmanager configuration
```

> Cortex Alertmanager can be uploaded via [Cortex Set Alertmanager API](../api/_index.md#set-alertmanager-configuration) or using Grafana Labs [Cortex Tools](https://github.com/grafana/cortex-tools).

Reviewer: ```suggestion
Cortex Alertmanager can be uploaded via Cortex [Set Alertmanager configuration API](../api/_index.md#set-alertmanager-configuration) or using Grafana Labs [Cortex Tools](https://github.com/grafana/cortex-tools).
```

Author: Done

Follow the instruction at the `cortextool` link above to download or update to the latest version of the tool.

To obtain the full help of how to use `cortextool` for all commands and flags, use
`cortextool --help-long`.

The following example shows the steps to upload the configuration to Cortex `Alertmanager` using `cortextool`. 

> 1. Create `.yaml` file to specify the configuration for Alertmanager notification characteristic.  

Reviewer: ```suggestion
#### 1. Create the Alertmanager configuration `.yml` file
```

The following is `amconfig.yml`, an example of a configuration for Cortex `Alertmanager` to send notification via email:

```
global:
  # The smarthost and SMTP sender used for mail notifications.
  smtp_smarthost: 'localhost:25'
  smtp_from: 'alertmanager@example.org'
  smtp_auth_username: 'alertmanager'
  smtp_auth_password: 'password'

route:
  # A default receiver.
  receiver: send-email

receivers:
  - name: send-email
    email_configs:
      - to: 'someone@localhost'
```

[Example on how to setup Slack](https://grafana.com/blog/2020/02/25/step-by-step-guide-to-setting-up-prometheus-alertmanager-with-slack-pagerduty-and-gmail/#:~:text=To%20set%20up%20alerting%20in,to%20receive%20notifications%20from%20Alertmanager.) to support receiving Alertmanager notification.

> 2. With your Cortex ID,  URL and API key,  you are ready to upload the configuration Alertmanager.

Reviewer: ```suggestion
#### 2. Upload the Alertmanager configuration
```

In this example,  Cortex `Alertmanager` is set to be available via localhost on port 8095 with only one user/org = 0.

To upload the above configuration `.yml` file with `--key` to be your Basic Authentication or API key:

```
cortextool alertmanager load ./amconfig.yml \
--address=http://localhost:8095 \
--id=0
--key=<yourKey>
```

> If there is no error reported,  the upload is successful.

Reviewer: ```suggestion
If there is no error reported, the upload is successful.
```

Author: @pracucci - you really pay attention to the space - lol

> 3. To confirm that the configuration is uploaded correctly:

Reviewer: ```suggestion
#### 3. Ensure the configuration has been uploaded correctly
```

```
cortextool alertmanager get \
--adress=http://localhost:8095 \
--id=0
--key=<yourKey>
```

> To upload the configuration for Cortex `Alertmanager` using Cortex API and curl - see [Cortex Set Alertmanager API](https://cortexmetrics.io/docs/api/#set-alertmanager-configuration).

Reviewer: This should be moved to the step (2) above.
